### ISSUE TITLE
Consider splitting SignInResult.NotAllowed in SignInResult.EmailConfirmationRequired and SignInResult.PhoneNumberConfirmationRequired

### ISSUE THREAD

**Issue by Ponant** [aspnet/Identity#1276]
`SignInResult.NotAllowed` is used in one place here
https://github.com/aspnet/Identity/blob/dev/src/Microsoft.AspNetCore.Identity/SignInManager.cs#L748 .
It does not distinguish between `IsEmailConfirmedAsync ` and `IsPhoneNumberConfirmedAsync`.
Distinguishing will make the code cleaner as it becomes more precise in returning errors to the dev and then the end UI user (see discussion here https://github.com/aspnet/Identity/issues/1265).
We could remove `SignInResult.NotAllowed` and do instead
`SignInResult.EmailConfirmationRequired`
and 
`SignInResult.PhoneNumberConfirmationRequired`
together with the correct Boolean flags to be caught in e.g. a Login Action.

I am aware that redefining the return type will propagate through methods, but it should not be too difficult. `PreSignInCheck` will become cleaner and would could profit from this to make `PasswordSigninAsync` cleaner as per the discussion in the referenced thread above. I can attempt a PR if you wish. Let me know, otherwise I will stop proposing changes. Cheers

**Comment by HaoK** [comment]
Feel free to file a PR for this and any related issues but we likely won't get to this until 2.1

**Comment by thomaslevesque** [comment]
This would be very helpful. `NotAllowed` tells us nothing about why sign-in failed.

**Comment by brockallen** [comment]
keep in mind that this would be a breaking change, as the semantics of the result values are now different.

**Comment by thomaslevesque** [comment]
@brockallen not necessarily. The new `EmailConfirmationRequired` result could still include the `IsNotAllowed` flag:

```
public static EmailConfirmationRequired { get; } = new SignInResult { IsNotAllowed = true, IsEmailConfirmationRequired = true };
```

So checking `IsNotAllowed` would still work.

**Comment by brockallen** [comment]
Oh you're proposing a new property to quality the meaning of the other?

**Comment by thomaslevesque** [comment]
> Oh you're proposing a new property to quality the meaning of the other?

Yes

**Comment by brockallen** [comment]
Then I'd almost want it to be a string -- something where I can put custom flags for my own apps. I have scenarios where I have 2fa onboarding, password resets, etc. that need to control workflow after first password successful authentication.
